File: webui/src/main/scala/com/taskadapter/web/event/EventBusImpl.scala
package com.taskadapter.web.event

import rx.lang.scala.subjects.{PublishSubject, SerializedSubject}
import rx.lang.scala.{Observable, Subscription}

import scala.reflect.ClassTag

object EventBusImpl extends EventBus {
  private val subject = PublishSubject[Event]()
  private val bus = SerializedSubject[Event](subject)

  def post(event: Event): Unit = {
    if (bus.hasObservers) {
      bus.onNext(event)
    }
  }

  def observable[T](eventClass: Class[T]): Observable[T] = {
    bus.filter(o => o != null) // Filter out null objects, better safe than sorry
      .filter(o => eventClass.isInstance(o)) // We're only interested in a specific event class
      .asInstanceOf[Observable[T]] // Cast it for easier usage
  }

  /**
    * Subscribe to the event, run the provided action when the event is received.
    *
    * @param f          the function to apply when an event of [T] class arrived
    * @tparam T type of the event to listen for
    * @return the subscription
    */
  def subscribe[T: ClassTag](f: T => Unit): Subscription = {
    val myClass = myClassOf[T]
    val subscription = observable(myClass).subscribe(e => f.apply(e))
    subscription
  }

  def myClassOf[T: ClassTag]: Class[T] = implicitly[ClassTag[T]].runtimeClass.asInstanceOf[Class[T]]
}

File: webui/src/main/java/com/taskadapter/web/event/EventBusImpl.java
package com.taskadapter.web.event;

import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

public class EventBusImpl {
    private static final Map<Class, List<Subscriber>> subscribers = new HashMap<>();

    public static void post(Event event) {
        if (!subscribers.containsKey(event.getClass())) {
            return;
        }
        subscribers.get(event.getClass())
                .forEach(subscriber -> subscriber.process(event));
    }

    /**
     * Subscribe to the event, run the provided action when the event is received.
     *
     * @param clazz event class to process
     * @param f     the function to apply when an event of [T] class arrived
     */
    public static synchronized <T extends Event> void subscribe(Class<T> clazz, Subscriber<T> f) {
        final List<Subscriber> list = EventBusImpl.subscribers.get(clazz);
        if (list == null) {
            var arrayList = new ArrayList<Subscriber>();
            arrayList.add(f);
            subscribers.put(clazz, arrayList);
        } else {
            list.add(f);
        }
    }
}
